sds_data_manager/lambda_code/SDSCode/pipeline_lambdas/batch_starter.py

```diff
@@ -354,6 +253,94 @@ def try_to_submit_job(session, job_info, start_date, version):
+        The data_source of the file that triggered the batch starter.
+    end_date : datetime, optional
+        The trigger file end date, by default None.
+
+    Returns
+    -------
+    None
+    """
+    # Find the files that this job depends on
+    dependency_event_msg = {
+        "data_source": job["data_source"],
+        "data_type": job["data_type"],
+        "descriptor": job["descriptor"],
+        "dependency_type": "UPSTREAM",
+        "relationship": "HARD",
+        "start_date": start_date,
+        "version": version,
+        "trigger_source": trigger_source,
+    }
+    if end_date:
+        dependency_event_msg["end_date"] = end_date
+
+    upstream_dependencies = _get_dependencies(dependency_event_msg)
+    if not upstream_dependencies.processing_input:
+        logger.info(
+            f"Upstream dependency not found for: {job['data_source']}, "
+            f"{job['data_type']}, "
+            f"{job['descriptor']}, "
+            f"{start_date}, "
+            f"{end_date}, "
+            f"{trigger_source}, "
+            f"{version}"
+        )
+        return  # Exit the loop early
+
+    logger.info(f"All dependencies found for the job: {job}")
+    # Find science processingInputs that have the same source as the potential job
+    for dep in upstream_dependencies.get_science_files():
+        if job["data_source"] == dep.source and isinstance(
+            dep, processing_input.ScienceInput
+        ):
+            # Try to start a downstream science job with the start_date from the
+            # upstream science dependency
+            # E.g.:
+            # if "job" == {"data_source":"swe","data_type":"l1b","descriptor":"sci"}
+            # And there is a processingInput in the upstream_dependencies that is
+            # {"type": "science",
+            #     "files": [
+            #         "imap_swe_l1a_sci_20240312_v001.cdf",
+            #         "imap_swe_l1a_sci_20240313_v001.cdf" ]}
+            # That means we need to kick off two swe l1b jobs for dates: "20240312" and
+            # "20240313"
+            for upstream_file in dep.imap_file_paths:
+                # TODO add function to processingInput to filter for start_date.
+                dep.imap_file_paths = [upstream_file]
+                dep.filename_list = [str(upstream_file.filename)]
+                job_start_date = datetime.strptime(upstream_file.start_date, "%Y%m%d")
+                job_version = upstream_file.version
+                # TODO when do we bump version?
+                try_to_submit_job(
+                    job, job_start_date, job_version, upstream_dependencies
+                )
+        # TODO should we break?
+        # For example if the potential job is mag_l1c (dependent on
+        # mag_l1b_norm_20250101 and mag_l1b_burst__20250101), we only want to
+        # kick off one mag l1c_20250101 job once and run it with both
+        # mag_l1b_norm and mag_l1b_burst as dependencies.
+        break
```

Comment:
Isn't this OK? We don't allow multiple jobs to be running at the same time, so one would fail and the other would succeed.